plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Completions.hs

```diff
@@ -0,0 +1,681 @@
+      pure []
+
+{- | Returns a list of all (and only) directories in the
+  directory described by path completion info
+-}
+listDirectoryCompletions :: Recorder (WithPriority Log) -> PathCompletionInfo -> IO [FilePath]
+listDirectoryCompletions recorder complInfo = do
+  filepaths <- listFileCompletions recorder complInfo
+  filterM (doesDirectoryExist . mkDirFromCWD complInfo) filepaths
+
+pathCompletionInfoFromCompletionContext :: CabalCompletionContext -> PathCompletionInfo
+pathCompletionInfoFromCompletionContext ctx =
+  PathCompletionInfo
+    { partialFileName = dirNamePrefix
+    , partialFileDir = Posix.addTrailingPathSeparator $ Posix.takeDirectory prefix
+    , cabalFileDir = dir
+    }
+ where
+  prefix = T.unpack $ completionPrefix ctx
+  dirNamePrefix = T.pack $ Posix.takeFileName prefix
+  dir = Posix.takeDirectory $ completionCabalFileDir ctx
+
+{- | Returns the directory, the currently handled cabal file is in.
+
+  We let System.FilePath handle the separator syntax since this is used
+  to query filepaths from the system. See Note [Using correct file path separators].
+-}
+mkCompletionDirectory :: PathCompletionInfo -> FilePath
+mkCompletionDirectory complInfo =
+  FP.addTrailingPathSeparator $
+    cabalFileDir complInfo FP.</> (FP.normalise $ partialFileDir complInfo)
+
+{- | Returns the complete filepath for the given filepath.
+
+  Since this is used for completions we use posix separators here.
+  See Note [Using correct file path separators].
+-}
+mkDirFromCWD :: PathCompletionInfo -> FilePath -> FilePath
+mkDirFromCWD complInfo fp =
+  Posix.addTrailingPathSeparator $
+    mkCompletionDirectory complInfo Posix.</> Posix.normalise fp
+
+{- | Information used to query and build file path/directory completions.
+
+  Note that partialFileName combined with partialFileDir results in
+  the original prefix.
+
+  Example:
+  On the written filepath: @dir1/fi@ the
+  resulting PathCompletionInfo would be:
+
+  @
+    partialFileName = "fi"
+    partialFileDir  = "dir1/dir2/fi"
+    ...
+  @
+-}
+data PathCompletionInfo = PathCompletionInfo
+  { partialFileName :: T.Text
+  -- ^ partly written start of next part of path
+  , partialFileDir  :: FilePath
+  -- ^ written part of path
+  , cabalFileDir    :: FilePath
+  -- ^ current working directory of the handled file
+  }
+  deriving (Eq, Show, Read)
+
+-- ----------------------------------------------------------------
+-- Completion Data
+-- ----------------------------------------------------------------
+
+-- | Keyword for cabal version; required to be the top line in a cabal file
+cabalVersionKeyword :: Map KeyWordName Completer
+cabalVersionKeyword =
+  Map.singleton "cabal-version:" $
+    constantCompleter $
+      map (T.pack . showCabalSpecVersion) [CabalSpecV2_2 .. maxBound]
+
+{- | Top level keywords of a cabal file.
+
+ TODO: we could add descriptions of field values and then show them when inside the field's context
```

Comment:
That would be pretty nice for discoverability

Reply:
Do you think it would be better to add it in this PR? I was planning to add this feature in a later PR.